[248] 멀티시그 상세 화면 변경

pio-noncelab opened the pull request:
# 주요 변경 사항

❗️지갑 상세 화면 파일 트리가 다수 변경되어 Conflict이 많이 발생할 수 있습니다. 1️⃣
❗️<백업하기 관련> 화면은 구현했지만 기능은 구현되어 있지 않습니다. 2️⃣

## 파일 트리 변경

1️⃣ 변경된 파일 트리
vault_menu/address_list_screen -> wallet_info/address_list_screen
vault_menu/info/multisig_export_options_screen -> wallet_info/export_options_screen
vault_menu/info/multisig_bsms_screen -> wallet_info/multisig_menu/coordinator_bsms_qr_screen
vault_menu/info/multisig_add_icon_bottom_sheet -> common/select_external_wallet_bottom_sheet
vault_menu/info/multisig_add_key_option_bottom_sheet -> wallet_info/multisig_menu/multisig_add_key_option_bottom_sheet
vault_menu/info/multisig_signer_memo_bottom_sheet -> wallet_info/multisig_menu/multisig_signer_name_bottom_sheet
vault_menu/info/multisig_setup_info_screen -> wallet_info/multisig_wallet_info_screen
vault_menu/info/name_and_icon_edit_bottom_sheet -> wallet_info/name_and_icon_edit_bottom_sheet
vault_menu/info/extended_pub_key_screen -> wallet_info/single_sig_menu/extended_pub_key_screen
vault_menu/info/mnemonic_view_screen -> wallet_info/single_sig_menu/mnemonic_view_screen
vault_menu/info/passphrase_check_screen -> wallet_info/single_sig_menu/passphrase_check_screen
vault_menu/info/passphrase_verification_screen -> wallet_info/single_sig_menu/passphrase_verification_screen
vault_menu/multisig_signer_bsms_export_screen -> wallet_info/single_sig_menu/signer_bsms_qr_screen
vault_menu/info/single_sig_setup_info_screen -> wallet_info/single_sig_wallet_info_screen
vault_menu/sync_to_wallet/sync_to_wallet_screen -> wallet_info/sync_to_wallet_screen
widgets/setup_info/setup_info_vault_item_card -> widgets/wallet_info/wallet_info_item_card
widgets/setup_info/setup_info_layout -> widgets/wallet_info/wallet_info_layout
providers/view_model/wallet_info/single_sig_wallet_info_view_model -> 

<table border="1" cellpadding="8" cellspacing="0">
  <tr>
    <th>변경 전 구조</th>
    <th>변경 후 구조</th>
  </tr>
  <tr>
    <td>
<pre>
vault_menu/
├── address_list_screen.dart
├── multisig_signer_bsms_export_screen.dart
├── sync_to_wallet/
│   └── sync_to_wallet_screen.dart
└── info/
    ├── multisig_export_options_screen.dart
    ├── multisig_bsms_screen.dart
    ├── multisig_add_icon_bottom_sheet.dart
    ├── multisig_add_key_option_bottom_sheet.dart
    ├── multisig_signer_memo_bottom_sheet.dart
    ├── multisig_setup_info_screen.dart
    ├── name_and_icon_edit_bottom_sheet.dart
    ├── extended_pub_key_screen.dart
    ├── mnemonic_view_screen.dart
    ├── passphrase_check_screen.dart
    ├── passphrase_verification_screen.dart
    └── single_sig_setup_info_screen.dart

widgets/
└── setup_info/
    ├── setup_info_vault_item_card.dart
    └── setup_info_layout.dart
</pre>
    </td>
    <td>
<pre>
wallet_info/
├── address_list_screen.dart
├── export_options_screen.dart
├── sync_to_wallet_screen.dart
│
├── multisig_wallet_info_screen.dart
├── single_sig_wallet_info_screen.dart
│
├── name_and_icon_edit_bottom_sheet.dart
├── common/
│   └── select_external_wallet_bottom_sheet.dart
│
├── multisig_menu/
│   ├── coordinator_bsms_qr_screen.dart
│   ├── multisig_add_key_option_bottom_sheet.dart
│   └── multisig_signer_name_bottom_sheet.dart
│
└── single_sig_menu/
    ├── extended_pub_key_screen.dart
    ├── mnemonic_view_screen.dart
    ├── passphrase_check_screen.dart
    ├── passphrase_verification_screen.dart
    └── signer_bsms_qr_screen.dart


widgets/
└── wallet_info/
    ├── wallet_info_item_card.dart
    └── wallet_info_layout.dart
       </pre>

  </tr>
</table>
- SingleSigInfoScreen, MultisigInfoScreen -> WalletInfoScreen 위젯 사용
- single_sig_wallet_info_view_model, multisig_wallet_info_view_model 삭제, wallet_info_view_model을 공통으로 사용

## 외부 지갑 선택 바텀 시트 공용 위젯으로 분리
- input
   - externalWalletButtonList: ExternalWalletButton (이름, 아이콘path) 리스트 (non-null)
   - selectedIndex: 선택된 인덱스(nullable)
- output
   - externalWalletButtonList의 요소를 1행당 최대 3개 까지 표시하여 바텀시트 화면으로 출력
- 예시
   - ``` dart
      final iconSourceList = [
      kCoconutVaultIconPath,
      kKeystoneIconPath,
      kSeedSignerIconPath,
      kJadeIconPath,
      kColdCardIconPath,
      kKruxIconPath,
      ];
      final externalWalletButtonList = [
        ExternalWalletButton(name: t.multi_sig_setting_screen.add_icon.coconut_vault, iconSource: iconSourceList[0]),
        ExternalWalletButton(name: t.multi_sig_setting_screen.add_icon.keystone3pro, iconSource: iconSourceList[1]),
        ExternalWalletButton(name: t.multi_sig_setting_screen.add_icon.seed_signer, iconSource: iconSourceList[2]),
        ExternalWalletButton(name: t.multi_sig_setting_screen.add_icon.jade, iconSource: iconSourceList[3]),
        ExternalWalletButton(name: t.multi_sig_setting_screen.add_icon.cold_card, iconSource: iconSourceList[4]),
        ExternalWalletButton(name: t.multi_sig_setting_screen.add_icon.krux, iconSource: iconSourceList[5]),
      ];
      final selectedIndex = iconSource != null ? iconSourceList.indexOf(iconSource) : null;
      final result = await MyBottomSheet.showDraggableBottomSheet<SignerSource?>(
        context: context,
        showDragHandle: false,
        maxChildSize: 0.45,
        minChildSize: 0.2,
        initialChildSize: 0.45,
        childBuilder: (context) => MultisigAddIconBottomSheet(iconSource: iconSource),
        childBuilder:
            (context) => SelectExternalWalletBottomSheet(
              externalWalletButtonList: externalWalletButtonList,
              selectedIndex: selectedIndex,
            ),
      );```


## 지갑 상세 화면

- 다중서명지갑 상세 화면 내 외부키 설정 변경
   - 외부 지갑의 + 아이콘 클릭 시 SignerSource 변경 바텀시트 호출
   - 외부 지갑 카드를 누르면 올라오던 **Memo 설정 바텀 시트**를 **이름 설정 바텀 시트**로 교체
   - **키 추가** 버튼을 통해 외부 지갑으로 등록됐던 시드를 내부 지갑으로 추가 할 수 있음
      - 니모닉 문구를 통한 추가: 복원하기 화면 재사용
      - SeedQR 스캔을 통한 추가: SeedQR 스캔 화면 재사용
      - -> 추가시 기존에 다른 지갑에도 해당 외부지갑으로 연결되어 있던 지갑이 내부 지갑으로 적용이 됨
- 지갑 상세 화면 내 메뉴 변경
   - 싱글 시그
      - 서명하기
      - 주소 보기
      - 확장 공개키 보기 (*추가)
      - 니모닉 문구 보기
      - 지갑 내보내기
         - 다중 서명키로 사용하기
         - 보기 전용 앱으로 내보내기
   - 멀티 시그
      - 다중 서명하기
      - 주소 보기
      - 지갑 내보내기
         - 다른 볼트와 공유하기
         - 보기 전용 앱으로 내보내기
         - 지갑 데이터 백업하기 (*추가, 화면만 구현됨)


## 🛠️ 테스트 항목
싱글시그, 멀티시그 변경된 화면 확인 및
**TS/TC 48-3번** 수행


### 이슈 번호
#248 

# 추가 작업 사항
제목 (Due time)
- [x] 지갑 이름 아이콘 수정 내 TextInput에 맞춰 메모 화면 수정(TextInput X버튼 위치, 글자 수 표기 스타일)
- [x] 키 추가 / 아이콘 버튼 바텀 시트 (X 버튼 위치, 글자 수 표기 스타일)
- [x] 싱글시그 상세 - 지갑 내보내기 - 다중 서명키로 사용하기의 하단 스타일 통일 (13:00)
- [x] 외부 지갑 메모 안에 있는 button 스타일이 달라보임 (13:20)
- [x] 다른 볼트와 공유하기 / 와치온리로 공유하기 툴팁 스타일이 달라보임 (13:40)
- [x] 스캔 화면(바탕이 어두운 화면)과 아닌 화면의 상단 패딩 조정 필요 (14:00)
- [x] Multisigner getSignerIconSource() 리팩토링(Map 형식) (14:40)
- [x] 외부 지갑 아이콘 설정 화면 팝업 문구 수정(14:50)
- [x] WalletIsolates verifyMnemonicMfp 비교시 uppercase(15:00)

pio-noncelab committed:
feat(multisig-signer-schema): MultisigSigner 데이터 모델 스키마 변경

- memo 삭제
- signerName 추가
- signerSource 추가

pio-noncelab committed:
feat(multisig-detail-screen): 다중서명지갑 상세 화면 내 외부키 설정 변경

- 아이콘 클릭 시 SignerSource 변경 가능
   - SignerSource 변경 시 안내 팝업 추가
- 외부 지갑 카드 영역 클릭 시 외부 지갑 이름 변경 가능
   - 이름이 설정되면 외부 지갑 카드 내 derivationPath 우측에 이름이 보임
- '키 추가' 버튼 클릭 시 키 '키 추가 옵션' 바텀시트 호출

pio-noncelab committed:
feat(multisig-signer-key): Signer로 사용되던 외부지갑에 니모닉 문구 입력을 통한 '키 추가' 기능 구현

- 니모닉 문구를 입력해 외부지갑 연동
   - 복원하기 화면을 이용하지만 버튼 클릭 시 해당 외부지갑의 MFP와 새로 추가되는 지갑의 MFP를 비교하는 로직이 추가됨
- 외부 지갑에 대해 이름이 저장되어 있는 경우, 이름/아이콘 설정 화면에서 이름 TextField에 자동으로 채워짐
- 이름/아이콘 설정을 끝내고 새로운 싱글시그 지갑이 추가 되면, 같은 MFP를 가진 지갑이 외부지갑으로 사용 되고 있는 경우 새로 만들어진 싱글시그 지갑으로 교체

pio-noncelab committed:
feat(multisig-signer-key): Signer로 사용되던 외부지갑에 SeedQR 스캔을 통한 '키 추가' 기능 구현

pio-noncelab committed:
feat(multisig-detail-menu): 멀티시그 지갑 상세 화면에서 메뉴 구성 변경

pio-noncelab committed:
feat(multisig-wallet-export): 멀티시그 지갑 내보내기 메뉴 구성

- 다른 볼트와 공유하기 상세화면
- 보기 전용 앱으로 내보내기 상세화면

pio-noncelab committed:
refactor(vault-setup-info-screen): 볼트 상세 화면 리팩토링 작업

SetupInfoLayout으로 통합
- 부모 클래스를 통해 상속받는 방식은, UI구조는 비슷하지만 위젯별 기능과 목적이 다르기 때문에 if/else로 분기가 많아지는 현상이 생기고, 추후 유지보수가 어려울 것으로 예상되어 사용하지 않음

pio-noncelab committed:
feat(single-sig-xpub-screen): 확장 공개키 보기 화면 추가

pio-noncelab committed:
refactor(file-tree): 지갑 상세 화면 파일 트리 변경

pio-noncelab committed:
refactor(wallet-info-screen): SingleSigInfoScreen, MultisigInfoScreen과 각각의 ViewModel을 WalletInfoScreen, WalletInfoViewModel로 통합 리팩토링 작업

pio-noncelab committed:
refactor(multisig-add-icon-bottom-sheet): 외부 기기 선택 바텀시트 분리

pio-noncelab committed:
feat(restore-memo-and-ui): 외부지갑 메모 기능 복원, 이름 변경 기능 제거, UI 수정

- 외부지갑 이름 변경 기능을 메모 기능으로 복원
- 외부지갑 아이콘 선택 바텀 시트 UI 수정
   - 외부 지갑 이름 부분 자간 높이 수정
   - 영문 문구 수정
   - '완료' 버튼을 누르는 시점에 변경사항이 적용되도록 수정
   - 메모 바텀 시트에 영문/일어 버전에서도 '닫기' 로 표시되는 부분 수정
   - 메모 바텀 시트 Title과 TextField 간격 조절
   - TextField 하단에 입력길이를 표시하는 부분에서 괄호 제거

doey-noncelab committed:
fix(ui): 더이상 사용하지 않는 문구 삭제, qr view 화면 툴팁 패딩 수정

pio-noncelab committed:
fix(multisig-signer-memo-bottomsheet): CupertinoTextField를 CoconutTextField로 교체

pio-noncelab committed:
fix(add-key-and-add-icon-bottomsheet): showDraggableScrollableSheet 위젯과 아이콘 입력 화면의 앱바에서 닫기 버튼과 title의 align이 어긋나는 현상 해결

ella-noncelab commented:
안드로이드 폰에서 멀티 시그 지갑의 '키 추가' 버튼 터치 시 카메라가 멈추는 에러가 발생

ella-noncelab committed:
fix(seed_qr_import): 안드로이드 카메라 에러 수정

ella-noncelab commented on lib/repository/wallet_repository.dart:
```diff
@@ -151,10 +151,25 @@ class WalletRepository {
+      String? importedDerivationPath;
+      try {
+        final bsms = Bsms.parseSigner(singlesigItem.signerBsms);
+        importedDerivationPath = bsms.signer?.path;
```
bsms.signer.path가 없는 경우가 있으면 안되어서, 이 경우에는 지갑 추가 에러 안내가 되어야 할 것 같아요.
```
importedDerivationPath = bsms.signer.path
그리고 현재 탭루트 지원하지 않으므로 무조건 p2wsh(Native SegWit multisig) derivationPath여야 하고 다른 path면 지갑 추가 될 수 없습니다.

pio-noncelab committed:
fix(signer-bsms-qr-screen): 싱글시그 상세 - 지갑 내보내기 - 다중서명키로 사용하기 화면 통일

pio-noncelab committed:
Merge branch 'feat/248-multisig-detail-ui' of https://github.com/noncelab/coconut_vault into feat/248-multisig-detail-ui

ella-noncelab commented on lib/screens/vault_creation/vault_name_and_icon_setup_screen.dart:
```diff
@@ -179,6 +204,27 @@ class _VaultNameAndIconSetupScreenState extends State<VaultNameAndIconSetupScree
+  /// externalSigner의 MFP를 사용해서 해당하는 멀티시그 지갑의 ID를 찾습니다.
+  int? _findMultisigVaultIdBySigner(MultisigSigner externalSigner) {
+    final signerMfp = externalSigner.keyStore.masterFingerprint;
+    final signerDerivationPath = externalSigner.getSignerDerivationPath();
```
이 조건만으로는 유일한 멀티시그지갑을 찾을 수 없는 것 같아요. 다시 한 번 확인 부탁드려요.
수정 중인 multisig 지갑 ID도 전달받는게 더 나을 것 같습니다.

doey-noncelab replied:
vault_name_and_icon_setup_screen.dart의 로직은 @roy-noncelab 할당받은 작업이랑 겹칠 것 같은데 중복 작업하지 않도록 해야할 것 같습니다.

pio-noncelab replied:
walletRepository의 _linkNewSinglesigVaultAndMultisigVaults() 메서드에서 새로 추가한 키를 외부지갑에 할당해 주기 때문에 같은 키를 외부 Signer로 사용하는 멀티시그지갑이 여러개일 때 모두 적용이 되고 있습니다.

doey-noncelab commented on lib/model/multisig/multisig_signer.dart:
```diff
@@ -43,4 +47,45 @@ class MultisigSigner {
+        break;
+      default:
+        iconPath = kAddCircleOutlinedIconPath;
+    }
```
`feat/247-add-singer-bsms` 브랜치 lib/enums/wallet_enums.dart에 HardwareType를 새롭게 정의했고 getter displayName도 정의했는데요. iconPath도 머지 후 model에서 빼는게 좋을 것 같습니다.

pio-noncelab committed:
fix(memo-bottom-sheet): 외부 지갑 메모 바텀 시트 UI 수정

- 앱바 적용('닫기' 버튼 대신 앱바의 X버튼 활용)
- '완료' 버튼은 FixedBottomButton으로 변경
- 글자 수 표기 스타일 VaultNameIconEditPalette와 통일

pio-noncelab committed:
fix(coordinator-bsms-tooltip): 다른 볼트와 공유하기 화면 툴팁 borderRadius 수정

pio-noncelab committed:
fix(custom-tooltip-top-padding): QR 스캐너 화면 처럼 뒷 배경이 어두운 경우 paddingTop을 20으로 설정

pio-noncelab committed:
refactor(signer-source-icon-map): MultisigSigner에서 case 문을 통해 가져오던 방식을 Map 형식으로 변경하고, enum으로 분리 작업 진행

pio-noncelab committed:
chore(slang): 아이콘 선택 바텀시트에서 선택 시 뜨는 팝업 문구 수정

pio-noncelab committed:
fix(verify-mnemonic-mfp): ActureMfp와 ExpectedMfp를 비교할 때 UpperCase로 변환 후 비교하도록 수정

ella-noncelab committed:
refactor(외부키추가): 함수 리팩토링

pio-noncelab committed:
fix(add-key): 키 추가 이후 키 추가한 시점에 있던 상세화면으로 다시 돌아오도록 수정

pio-noncelab committed:
fix(multisig-add-key-bottomsheet): 바텀시트 반환 값을 통해 Navigator 처리하도록 수정

pio-noncelab committed:
fix(vault-name-and-icon-setup): multisigVaultId가 null이어도 의도하지 않는 동작이 발생하지 않도록 수정

pio-noncelab committed:
style(rename): walletCreationProvider에 추가했던 변수(_callBackFromVaultId) 리네이밍

ella-noncelab committed:
refactor(renaming): callBackFromVaultId -> multisigVaultIdOfExternalSigner

ella-noncelab committed:
Merge branch 'refactor/248-multisig-detail' into feat/248-multisig-detail-ui
